src/SignalingClient.ts

```diff
@@ -321,15 +359,36 @@ export class SignalingClient extends EventEmitter {
      * 'error' event handler. Forwards the error onto listeners.
      */
     private onError(error: Error | Event): void {
+        console.log("On error invoked");
         this.emit('error', error);
     }
 
     /**
      * 'close' event handler. Forwards the error onto listeners and cleans up the connection.
      */
-    private onClose(): void {
+    private onClose(event?: CloseEvent): void {
         this.readyState = ReadyState.CLOSED;
-        this.cleanupWebSocket();
-        this.emit('close');
+        console.log("On close invoked");
+        switch (event.code) {
+            // Going away
+            case 1001:
+            // Abnormal closure
+            case 1006:
+            // Service Restart
+            case 1012:
+            // Try Again Later
+            case 1013:
+            // TLS Handshake
+            case 1015:
+                console.log("Allowing reconnect if option enabled on test page for code", event.code);
+                this.cleanupWebSocket();
+                this.emit('closewithretry');
+                break;
+            default:
+                console.log("No reconnect will be attempted. Just exiting with code", event.code);
+                this.cleanupWebSocket();
+                this.emit('close');
```

Comment:
Can we also emit the `event` as data? Then someone can take this data and take action
```suggestion
                this.emit('close', event);
```